LDpred2: Append score to existing output file

deepchocolate opened the pull request:
<!-- To ensure we can review your pull request promptly please complete this template entirely. -->

<!-- Please reference the issue number here. You can replace "Fixes" with "Closes" if it makes more sense. -->
Fixes #144 
Fixes #151 
Fixes #153 

Changes proposed in this pull request:
<!-- Please list all changes/additions here. -->
- Added the possibility to append a score to an existing output file using the option `--out-merge` and optionally `--out-merge-ids` (if ID columns are not IID and FID).
- Moved code into functions
- Fixed error when character values are present in the chromosome column (filtered out)
- Added tests to cover this and existing functionality
- Added missing sumstat snippets to repo

I note that it is possible that this functionality may raise conflicts if the user runs several scorings in parallel, and both processes read the existing data just prior to the writing of a parallell process, but I'm guessing the chance of that happening is low enought to outweigh the benefits of this functionality.

## Before submitting

<!-- Please complete this checklist BEFORE submitting your PR to speed along the review process. -->
- [x] I've read and followed all steps in the [Making a pull request](https://github.com/comorment/containers/blob/main/CONTRIBUTING.md#making-a-pull-request)
    section of the `CONTRIBUTING` docs.
- [x] I've updated or added any relevant docstrings following the syntax described in the
    [Writing docstrings](https://github.com/comorment/containers/blob/main/CONTRIBUTING.md#writing-docstrings) section of the `CONTRIBUTING` docs.
- [ ] If this PR fixes a bug, I've added a test that will fail without my fix.
- [x] If this PR adds a new feature, I've added tests that sufficiently cover my new functionality.

deepchocolate commented:
Added clarification of documentation, and an incorrect error message in `complementSumstats()`

espenhgn commented on usecases/LDpred2_test/run.sh:
```diff
@@ -14,17 +21,28 @@ export DIR_REFERENCE=$DIR_BASE/reference
 
 # Tutorial data
-# Phenotypic data is part of bed file
+# Phenotypic data is part of public-data3 fam file
```
public-data3.fam file (missing punctuation)

espenhgn commented on usecases/LDpred2_test/run.sh:
The test script now crashes because of a missing file (`sumstats.txt`):
```
$ bash run.sh 
### Running R function unittests
Test passed 
Test passed 
Test passed 
Error in data.table::fread(input, ..., data.table = data.table, nThread = nThread) : 
  File '/nrec/space/espenh/containers/usecases/LDpred2_test/unittest/data/sumstats.txt' does not exist or is non-readable. getwd()=='/nrec/space/espenh/containers/usecases/LDpred2_test'
Calls: <Anonymous> -> <Anonymous>
Execution halted
### Testing sumstats scripts
Error in data.table::fread(input, ..., data.table = data.table, nThread = nThread) : 
  File '/nrec/space/espenh/containers/usecases/LDpred2_test/unittest/data/sumstats.txt' does not exist or is non-readable. getwd()=='/nrec/space/espenh/containers/usecases/LDpred2_test'
Calls: <Anonymous> -> <Anonymous>
Execution halted
```

deepchocolate replied:
Those files were left out due to the gitignore, but I've added them now

espenhgn reviewed (changes requested):
Thanks! Looks good, but some file is missing with the test script (see above), however.

espenhgn reviewed (changes requested):
Thanks @deepchocolate; 
I encountered another issue with a missing file:
```
### Testing ldpred2.R (various options, manually downloaded LD, output merge)
Downloading LD map
--2023-04-11 07:25:29--  https://figshare.com/ndownloader/files/37802721
Resolving figshare.com (figshare.com)... 2a05:d018:1f4:d000:e143:4beb:37ea:9af4, 2a05:d018:1f4:d003:c91e:f786:d93e:5593, 34.248.204.94, ...
Connecting to figshare.com (figshare.com)|2a05:d018:1f4:d000:e143:4beb:37ea:9af4|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/37802721/map_hm3_plus.rds?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230411/eu-west-1/s3/aws4_request&X-Amz-Date=20230411T072529Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=eec81eeb428c5ed9eeb0e896f415e27472906810f9a5364e86ab08fa01133a65 [following]
--2023-04-11 07:25:29--  https://s3-eu-west-1.amazonaws.com/pfigshare-u-files/37802721/map_hm3_plus.rds?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIYCQYOYV5JSSROOA/20230411/eu-west-1/s3/aws4_request&X-Amz-Date=20230411T072529Z&X-Amz-Expires=10&X-Amz-SignedHeaders=host&X-Amz-Signature=eec81eeb428c5ed9eeb0e896f415e27472906810f9a5364e86ab08fa01133a65
Resolving s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)... 52.218.37.43, 52.218.26.107, 52.92.16.80, ...
Connecting to s3-eu-west-1.amazonaws.com (s3-eu-west-1.amazonaws.com)|52.218.37.43|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 43814795 (42M) [application/octet-stream]
Saving to: '/nrec/space/espenh/containers/usecases/LDpred2_test/data/ld/map_hm3_plus.rds'

/nrec/space/espenh/containers/usecases/LDpred2_t 100%[=========================================================================================================>]  41.78M  51.0MB/s    in 0.8s    

2023-04-11 07:25:30 (51.0 MB/s) - '/nrec/space/espenh/containers/usecases/LDpred2_test/data/ld/map_hm3_plus.rds' saved [43814795/43814795]

Test error: Character in chromosome column
Loading backingfile: /nrec/space/espenh/containers/usecases/LDpred2_test/data/EUR_imputed.rds 

### Reading LD reference meta-file from  /nrec/space/espenh/containers/usecases/LDpred2_test/data/ld/ldref/map.rds 
Error in gzfile(file, "rb") : cannot open the connection
Calls: readRDS -> gzfile
In addition: Warning message:
In gzfile(file, "rb") :
  cannot open compressed file '/nrec/space/espenh/containers/usecases/LDpred2_test/data/ld/ldref/map.rds', probable reason 'No such file or directory'
Execution halted
```
Are you pointing to the correct file, or did it fail to process/download?

espenhgn commented:
> @espenhgn It's that the script downloads the `map_hm3_plus.rds` which is actually not used in the scripts. I think that's because I didn't find any file with LD blocks for this map. So I kept working with the old maps (`map.rds`), but didn't notice the error as that file was always available in my development environment. But I've added the download of the old map for now so it should solve your error.

Hi again. Sorry, the file is still missing. If you have issues due to the .gitignore file it may be forcefully added using the syntax `git add --force my/ignore/file.foo`. Hope that works :)

deepchocolate commented:
> > @espenhgn It's that the script downloads the `map_hm3_plus.rds` which is actually not used in the scripts. I think that's because I didn't find any file with LD blocks for this map. So I kept working with the old maps (`map.rds`), but didn't notice the error as that file was always available in my development environment. But I've added the download of the old map for now so it should solve your error.
> 
> Hi again. Sorry, the file is still missing. If you have issues due to the .gitignore file it may be forcefully added using the syntax `git add --force my/ignore/file.foo`. Hope that works :)

Hopefully fixed now. I've removed downloading of maps, and used maps inside the lpred2_ref repository. However, the tests now assume that this repository is available in the parent directory of this repository. I forgot that the unittests for the tutorial data actually doesn't use these files, because it calculates LD instead so those tests are unaffected.

espenhgn reviewed (commented):
Hi again, there is now another test that fails on my end:
```
Test appending score to output file
Runing unittests on output
-- Failure (???): Appended score output ----------------------------------------
hasAllColumns(tmp, cnamesExp) is not TRUE

`actual`:   FALSE
`expected`: TRUE 

Error in reporter$stop_if_needed() : Test failed
Calls: test_that -> <Anonymous>
Execution halted
```

deepchocolate commented:
> Hi again, there is now another test that fails on my end:
> 
> ```
> Test appending score to output file
> Runing unittests on output
> -- Failure (???): Appended score output ----------------------------------------
> hasAllColumns(tmp, cnamesExp) is not TRUE
> 
> `actual`:   FALSE
> `expected`: TRUE 
> 
> Error in reporter$stop_if_needed() : Test failed
> Calls: test_that -> <Anonymous>
> Execution halted
> ```

Quite "funny", I had the same test fail. In my case it was the run of lpred2 prior to that test that failed without throwing the right error code. When I inspected it seemed that the process was terminated by the Ubuntu (22) in some way, likely due to a memory issue. Chromium also crashed simultaneously. 

Can you do an echo of `dump` to see if there's anything weird happening?
https://github.com/comorment/containers/blob/1bc3a659ef49681c6fe68492f823ae0dc3d2c359/usecases/LDpred2_test/scripts/extended.sh#L52

espenhgn commented:
Lack of memory could be a culprit. The "devbox" environment I use only have 8GB of RAM. I'll check `dump`.

espenhgn commented on usecases/LDpred2_test/scripts/extended.sh:
```diff
@@ -0,0 +1,57 @@
+done
+echo "Test appending score to output file"
+export fileExisting=$fileOut.inf
+export scoreNameExisting=score # Exists from runs above
```
@deepchocolate; I think the files `score` and `newScore` on the next line are not correctly defined. Either these should be explicitly defined (`score.txt`) or via some environment variable (`$score`). The `extended.R` script won't now load the correct files, causing the test crash.

deepchocolate replied:
Not sure I follow... `$fileOut` is defined by `run.sh` to "public-data.score" which takes on a .inf or .auto extension in this place..

The only way I can replicate this error is by starting up Chrome otherwise it runs fine

espenhgn replied:
Hi. It's wasn't `$fileOut.inf` that I referred to. It was `score` and `newScore` on line 50-51. I now understand they are column names; however, the column name `newScore` is not found in the output file in the test. 
I added 
```
  print(c(fileExisting, scoreNameExisting, scoreNameNew))
  print(cnames)
  print(cnamesExp)
```
to the `test_that` function in `extended.R` script, yielding
```
[1] "/nrec/space/espenh/containers/usecases/LDpred2_test/output/public-data.score.inf"
[2] "score"                                                                           
[3] "newScore"                                                                        
[1] "FID"   "IID"   "score"
[1] "FID"      "IID"      "score"    "newScore"
```
so the output file is not merged as intended. 
`echo "$dump"` yielded
```
Test appending score to output file
Checking ability to merge score with file /nrec/space/espenh/containers/usecases/LDpred2_test/output/public-data.score.inf due to --out-merge
Loading backingfile: /nrec/space/espenh/containers/usecases/LDpred2_test/data/public-data3.rds 

### Reading LD reference meta-file from  /ldpred2_ref/map_hm3_plus.rds 

### Reading summary statistics /nrec/space/espenh/containers/usecases/LDpred2_test/data/tutorial_data/public-data3-sumstats.txt 
Loaded 50000 SNPs
Filtering SNPs using /REF/hapmap3/w_hm3.justrs 
Retained 49085 out of 50000 
Filtering SNPs based on --chr2use
Retained 49085 out of 49085 
Matching sumstats to genotypes
49,085 variants to be matched.
0 ambiguous SNPs have been removed.
45,009 variants have been matched; 22,591 were flipped and 30,040 were reversed.
Matching sumstats to LD reference
45,009 variants to be matched.
0 ambiguous SNPs have been removed.
44,553 variants have been matched; 0 were flipped and 0 were reversed.
Warning message:
In merge.data.table(as.data.table(sumstats4), as.data.table(info_snp),  :
  column names 'pos.ss' are duplicated in the result

### Loading LD reference from  /ldpred2_ref/ldref_hm3_plus/LD_with_blocks_chr@.rds 
         LD_with_blocks_chr1.rds : loading LD for 3495 out of 117534 SNPs
Runing unittests on output
```
I'm not sure if the Warning shown here is related.

What does Chrome (the browser?) have to do with this, I wonder?

deepchocolate replied:
@espenhgn  As you can see in the dump 
```
### Loading LD reference from  /ldpred2_ref/ldref_hm3_plus/LD_with_blocks_chr@.rds 
         LD_with_blocks_chr1.rds : loading LD for 3495 out of 117534 SNPs
```
it stops at chromosome 1, so there is an error occuring there. If you have a similar issue as me, the error is not provided in the output (`$dump`) because the process is killed by the OS.
To see the error you have to replace this:
https://github.com/comorment/containers/blob/1bc3a659ef49681c6fe68492f823ae0dc3d2c359/usecases/LDpred2_test/scripts/extended.sh#L52

With `$LDP --ldpred-mode inf --name-score $scoreNameNew --out $fileExisting --out-merge`, i.e. don't assign the output to a variable.

I have no idea why chrome causes this error for me, but it must be something with the memory. I noted that Rstudio seem to cause the same problem. I think Rstudio uses QtWebEnginge (https://wiki.qt.io/QtWebEngine) which relies on chromium. So if you're runnning Rstudio while running the test I guess it could cause the same issue as for me

deepchocolate replied:
> Hi. I'm not using Rstudio for this testing, just `bash` in a virtual machine setup. It has 8GB of RAM, which has also caused some memory issues with loading the reference files in some other example(s). I'm not sure if there are any alternatives to [`readRDS`](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/readRDS) that allow reading only a subset of the file in R, i.e., the elements `[ind.chr3, ind.chr3]`.

I don't use RStudio for testing either, but it's enough for me that Rstudio or Chrome is running for this problem to appear

espenhgn replied:
Ok, if you are confident this code runs with enough memory available, I'm fine with merging this as is of course. 
I suppose the chromium-based software like to buffer a lot of memory even if it is not actually used.

deepchocolate replied:
Yes, I'm confident it works.. but I will test to see if some complete scoring runs can be dropped or maybe downsize the datasets used. I think the tests of the tutorial results are most important to run in full

espenhgn replied:
Ok, sounds reasonable. I just checked; simply reading in `/ldpred2_ref/ldref_hm3_plus/LD_with_blocks_chr1.rds` in R consumes about 1.9GB of system memory, so if there are options to "free" memory when data is no longer needed by the scripts it may help (`rm(foo); gc()` or something).

deepchocolate replied:
Wow, I didn't realize it was that hungry! Good idea! I'll see if that can be implemented as well
